### PR TITLE
feat: add initial support for contexts

### PR DESCRIPTION
**Pull request by lstocchi**
### What does this PR do?

This adds an initial support to work with contexts.
It has been extracted from https://github.com/containers/podman-desktop/pull/3131 to allow a quicker and easier review/merge.

### Screenshot/screencast of this PR

N/A

### What issues does this PR fix or reference?

it is part of https://github.com/containers/podman-desktop/pull/3131

### How to test this PR?

N/A

@benoitf this also contain the fixes based on your comments in https://github.com/containers/podman-desktop/pull/3131

Still not sure what text i should add to mark that some code has been taken from the Vscode project

### CODE REVIEW

**Comment by benoitf** [comment]
@lstocchi include copyright and some links as for example https://github.com/eclipse-theia/theia/blob/41c6d9f6f15caf9b2059b6a03c91699d31d80029/packages/plugin-ext/src/common/uint.ts#L17

File: packages/main/src/plugin/context-registry.ts [hunk 1]
```diff
@@ -28,13 +28,21 @@ export class ContextRegistry {
...
-      const newId = ctxWithLargestId.id + 1;
-      const ctx = new Context(newId, null, extension, this.apiSender);
+      const id = this.getNewId();
+      const ctx = new Context(id, null, extension, this.apiSender);
       this.contexts.push(ctx);
     }
   }
 
+  getNewId() {
+    let newId = 0;
```
Comment (benoitf) [inline packages/main/src/plugin/context-registry.ts]: just a question : do we need to have this code or we could just increment the counter ?

File: packages/main/src/plugin/api/context-info.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  | Array<null | undefined | boolean | number | string>
+  | Record<string, null | undefined | boolean | number | string>;
+
+export interface IContext {
+  getValue<T extends ContextKeyValue = ContextKeyValue>(key: string): T | undefined;
+}
+
+export interface ContextInfo {
+  readonly id: number;
+  readonly parent: Context | null;
```
Comment (benoitf) [inline packages/main/src/plugin/api/context-info.ts]: I think we almost never use null, always undefined so do we have strong interest in having null and not undefined for these two properties?

File: packages/main/src/plugin/context/context.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import type { ApiSenderType } from '../api.js';
+import type { IContext } from '../api/context-info.js';
+
+export class Context implements IContext {
+  private _value: Record<string, any>;
+
+  constructor(
+    private _id: number,
+    private _parent: Context | null,
```
Comment (benoitf) [inline packages/main/src/plugin/context/context.ts]: from my understanding parent and extension are optional so they should be at the end of the parameters

also probably same than before, probably it's optional so undefined and not null

File: packages/renderer/src/stores/contexts.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+export const contexts: Writable<ContextUI[]> = writable([]);
+
+// use helper here as window methods are initialized after the store in tests
+const listContexts = async (): Promise<ContextUI[]> => {
+  // retrieve all contexts saved on the server side
+  const serverContexts = await window.listContexts();
+  // as we cannot send an instance from main to renderer, we create a clone with the contextUI type so its functions can be invoked
+  const adaptedContexts: ContextUI[] = [];
+  serverContexts.forEach(ctx => {
+    let parentCtx: ContextUI | null = null;
```
Comment (benoitf) [inline packages/renderer/src/stores/contexts.ts]: we should not use 'null'

File: packages/main/src/plugin/context-registry.ts (hunk 1 above)
Comment (lstocchi) [reply packages/main/src/plugin/context-registry.ts]: no you're right, done!!

File: packages/main/src/plugin/api/context-info.ts (hunk 2 above)
Comment (lstocchi) [reply packages/main/src/plugin/api/context-info.ts]: no strong opinion, removed for the moment.

File: packages/main/src/plugin/context-registry.ts (hunk 1 above)
Comment (benoitf) [reply packages/main/src/plugin/context-registry.ts]: it was just a question, but yes if we can keep things simple

**Review by benoitf — APPROVED** [review]

File: packages/renderer/src/lib/context/context.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+  constructor(private _id: number, private _parent?: ContextUI, private _extension?: string) {
+    this._value = {};
+  }
+
+  get id(): number {
+    return this._id;
+  }
+
+  get extension(): string | null {
```
Comment (benoitf) [inline packages/renderer/src/lib/context/context.ts]: it looks we still use null there but it should be undefined no ?

Comment (lstocchi) [reply packages/renderer/src/lib/context/context.ts]: I missed it. Fixed!!

**Review by benoitf — APPROVED** [review]

**Pull request merged**